operator/builtin/input/file/reader.go

```diff
@@ -106,6 +150,12 @@ func (f *Reader) ReadToEnd(ctx context.Context) {
 	fr := NewFingerprintUpdatingReader(f.file, f.Offset, f.Fingerprint, f.fileInput.fingerprintSize)
 	scanner := NewPositionalScanner(fr, f.fileInput.MaxLogSize, f.Offset, f.fileInput.SplitFunc)
 
+	if f.fileInput.labelRegex != nil {
```

Comment:
This conditional block will execute during every polling cycle, rather than just the first time the file is seen.

I think we should move this call into `file.go`, in the `newReader` method, after we've instantiated a new reader, but before we call `InitializeOffset`. At this point, we know that this is the first time we've seen this file, so it would be the appropriate time to consume the headers. We can then set the initial offset appropriately as well. (if startAtBeginning, set it according to how many bytes were consumed for headers)

Reply:
I think this would work:

```go
func (f *InputOperator) newReader(file *os.File, fp *Fingerprint, firstCheck bool) (*Reader, error) {
	// Check if the new path has the same fingerprint as an old path
	if oldReader, ok := f.findFingerprintMatch(fp); ok {
		newReader, err := oldReader.Copy(file)
		if err != nil {
			return nil, err
		}
		newReader.fileLabels = f.resolveFileLabels(file.Name())
		return newReader, nil
	}

	// If we don't match any previously known files, create a new reader from scratch
	newReader, err := f.NewReader(file.Name(), file, fp)
	if err != nil {
		return nil, err
	}
	if f.fileInput.labelRegex != nil {
		newReader.readHeaders()
	}
	startAtBeginning := !firstCheck || f.startAtBeginning
	if err := newReader.InitializeOffset(startAtBeginning); err != nil {
		return nil, fmt.Errorf("initialize offset: %s", err)
	}
	return newReader, nil
}
```

Reply:
Good idea, done.